File: code/go/ch10/src/luago/vm/inst_upvalue.go
package vm

import . "luago/api"

// R(A) := UpValue[B]
func getUpval(i Instruction, vm LuaVM) {
	a, b, _ := i.ABC()
	a += 1
	b += 1

	vm.Copy(LuaUpvalueIndex(b), a)
}

// UpValue[B] := R(A)
func setUpval(i Instruction, vm LuaVM) {
	a, b, _ := i.ABC()
	a += 1
	b += 1

	vm.Copy(a, LuaUpvalueIndex(b))
}

// R(A) := UpValue[B][RK(C)]
func getTabUp(i Instruction, vm LuaVM) {
	a, b, c := i.ABC()
	a += 1
	b += 1

	vm.GetRK(c)
	vm.GetTable(LuaUpvalueIndex(b))
	vm.Replace(a)
}

// UpValue[A][RK(B)] := RK(C)
func setTabUp(i Instruction, vm LuaVM) {
	a, b, c := i.ABC()
	a += 1

	vm.GetRK(b)
	vm.GetRK(c)
	vm.SetTable(LuaUpvalueIndex(a))
}

File: code/csharp/ch13/luavm/vm/InstUpvalue.cs
using ArithOp = System.Int32;
using CompareOp = System.Int32;
using LuaVM = luavm.state.LuaState;

namespace luavm.vm
{
    public class InstUpvalue
    {
        internal static void GetUpval(Instruction i, ref LuaVM vm)
        {
            var (a, b, _) = i.ABC();
            a += 1;
            b += 1;

            vm.Copy(LuaVM.LuaUpvalueIndex(b), a);
        }


        // UpValue[B] := R(A)
        internal static void SetUpval(Instruction i, ref LuaVM vm)
        {
            var (a, b, _) = i.ABC();
            a += 1;
            b += 1;

            vm.Copy(a, LuaVM.LuaUpvalueIndex(b));
        }

        // R(A) := UpValue[B][RK(C)]
        internal static void GetTabUp(Instruction i, ref LuaVM vm)
        {
            var (a, b, c ) = i.ABC();
            a += 1;
            b += 1;

            vm.GetRK(c);
            vm.GetTable(LuaVM.LuaUpvalueIndex(b));
            vm.Replace(a);
        }

        // UpValue[A][RK(B)] := RK(C)
        internal static void SetTabUp(Instruction i, ref LuaVM vm)
        {
            var ( a, b, c) = i.ABC();
            a += 1;

            vm.GetRK(b);
            vm.GetRK(c);
            vm.SetTable(LuaVM.LuaUpvalueIndex(a));
        }
    }
}